### PR TITLE
Abstract points calculations into separate classes

### PR DESCRIPTION
**Pull request by muyjohno**
This allowed me to add weekly cap to participation points. Refactored a few things too.

Some values changed in the tests because I swapped a result around in the sample league.

### CODE REVIEW

File: app/models/points/participation.rb [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    def calculate(game, row)
+      @game = game
+      @row = row
+
+      points_for_participation.tap do |points|
+        return 0 if current >= max || week_current >= week_max
+        return (max - current) if (current + points) >= max
+        return (week_max - week_current) if (week_current + points) >= week_max
+      end
+    end
```
Comment (richdouglasevans) [inline app/models/points/participation.rb]: Huh, `#tap` is cute. Not seen that before outside of JavaScript and it's not baked into the language like this is. (You could write it yourself easily enough or use a library such as LoDash.)

:+1: for showing me something.

File: spec/models/leaderboard_spec.rb [hunk 2]
```diff
@@ -28,6 +28,15 @@
       it { expect(row.games).not_to include(game5) }
     end
 
+    describe("#games_for_week") do
+      let(:week2_game) { create(:game, corp: adam, runner: ben, week: 2) }
+      let(:leaderboard) { create_leaderboard(games: [game1, game2, week2_game]) }
```
Comment (richdouglasevans) [inline spec/models/leaderboard_spec.rb]: You monster Johno. Have you no shame?

:kiss: :laughing:

File: spec/models/points/result_spec.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  context "bye" do
+    before do
+      create(:rule, key: :points_for_bye, value: 9)
+    end
+
+    let(:game) { create(:game, corp: adam, runner: nil, result: :bye) }
+
+    it { expect(subject.calculate(game, adam_row)).to eq(9) }
+  end
+end
```
Comment (richdouglasevans) [inline spec/models/points/result_spec.rb]: Aw, Adam, Ben, Chuck => A, B, C. I tend to use the names of friends and family.

(With apologies if you have friends and family called Adam, Ben, and Chuck.)

:smile:

File: app/models/points/participation.rb (hunk 1 above)
Comment (muyjohno) [inline app/models/points/participation.rb]: If anything, I overuse `tap`! I have a passionate dislike of dangling return variables, so
```
def do_the_thing
  thing = whatever
  thing.do_it!
  thing
end
```
reads much more nicely to me as:
```
def do_the_thing
  whatever.tap do |thing|
    thing.do_it!
  end
end
```

**Pull request merged**